File: Striver/DP/Q52.py
# evalute boolean expression to true

# recursive solution
def countWays(s, i, j, isTrue):
    if i > j: return 0
    if i == j:
        if isTrue:
            return 1 if s[i] == 'T' else 0
        else:
            return 1 if s[i] == 'F' else 0

    ans = 0
    for k in range(i+1, j, 2):
        lt = countWays(s, i, k-1, True)
        lf = countWays(s, i, k-1, False)
        rt = countWays(s, k+1, j, True)
        rf = countWays(s, k+1, j, False)

        if s[k] == '&':
            if isTrue:
                ans += lt*rt
            else:
                ans += lt*rf + lf*rt + lf*rf
        elif s[k] == '|':
            if isTrue:
                ans += lt*rt + lt*rf + lf*rt
            else:
                ans += lf*rf
        elif s[k] == '^':
            if isTrue:
                ans += lt*rf + lf*rt
            else:
                ans += lt*rt + lf*rf

    return ans

# memoization solution
def countWaysMemo(s, i, j, isTrue, memo):
    if i > j: return 0
    if i == j:
        if isTrue:
            return 1 if s[i] == 'T' else 0
        else:
            return 1 if s[i] == 'F' else 0

    if memo[i][j][isTrue] != -1: return memo[i][j][isTrue]

    ans = 0
    for k in range(i+1, j, 2):
        lt = countWaysMemo(s, i, k-1, True, memo)
        lf = countWaysMemo(s, i, k-1, False, memo)
        rt = countWaysMemo(s, k+1, j, True, memo)
        rf = countWaysMemo(s, k+1, j, False, memo)

        if s[k] == '&':
            if isTrue:
                ans += lt*rt
            else:
                ans += lt*rf + lf*rt + lf*rf
        elif s[k] == '|':
            if isTrue:
                ans += lt*rt + lt*rf + lf*rt
            else:
                ans += lf*rf
        elif s[k] == '^':
            if isTrue:
                ans += lt*rf + lf*rt
            else:
                ans += lt*rt + lf*rf

    memo[i][j][isTrue] = ans
    return ans

# tabulation solution
def countWaysTab(s):
    n = len(s)
    dp = [[[0]*2 for _ in range(n)] for _ in range(n)]

    for i in range(n-1,-1,-1):
        for j in range(i,n):
            if i > j: continue
            if i == j:
                dp[i][j][0] = 1 if s[i] == 'F' else 0
                dp[i][j][1] = 1 if s[i] == 'T' else 0
                continue

            for k in range(i+1, j, 2):
                lt = dp[i][k-1][1]
                lf = dp[i][k-1][0]
                rt = dp[k+1][j][1]
                rf = dp[k+1][j][0]

                if s[k] == '&':
                    dp[i][j][1] += lt*rt
                    dp[i][j][0] += lt*rf + lf*rt + lf*rf
                elif s[k] == '|':
                    dp[i][j][1] += lt*rt + lt*rf + lf*rt
                    dp[i][j][0] += lf*rf
                elif s[k] == '^':
                    dp[i][j][1] += lt*rf + lf*rt
                    dp[i][j][0] += lt*rt + lf*rf
    return dp[0][n-1][1]

s = "T|T&F^T"
n = len(s)
memo = [[[-1]*2 for _ in range(n)] for _ in range(n)]
print(countWays(s, 0, n-1, True))
print(countWaysMemo(s, 0, n-1, True, memo))
print(countWaysTab(s))

File: Striver/DP/Q52.java
package Striver.DP;

public class Q52 {
    // evaluate boolean expression to true

    // recursive solution - Time: O(3^n) Space: O(n)
    public int countWays(String s, int i, int j, boolean isTrue) {
        if(i > j) return 0;
        if(i == j){
            if(isTrue) return s.charAt(i) == 'T' ? 1 : 0;
            else return s.charAt(i) == 'F' ? 1 : 0;
        }

        int ans = 0;
        for(int k = i+1; k <= j-1; k+=2){
            int lt = countWays(s, i, k-1, true);
            int lf = countWays(s, i, k-1, false);
            int rt = countWays(s, k+1, j, true);
            int rf = countWays(s, k+1, j, false);

            if(s.charAt(k) == '&'){
                if(isTrue) ans += lt * rt;
                else ans += lt * rf + lf * rt + lf * rf;
            }
            else if(s.charAt(k) == '|'){
                if(isTrue) ans += lt * rt + lt * rf + lf * rt;
                else ans += lf * rf;
            }
            else if(s.charAt(k) == '^'){
                if(isTrue) ans += lt * rf + lf * rt;
                else ans += lt * rt + lf * rf;
            }
        }

        return ans;
    }

    // memoization - Time: O(n^3) Space: O(n^2)
    public int countWaysMemo(String s, int i, int j, boolean isTrue, int[][][] dp) {
        if(i > j) return 0;
        if(i == j){
            if(isTrue) return s.charAt(i) == 'T' ? 1 : 0;
            else return s.charAt(i) == 'F' ? 1 : 0;
        }
        if(dp[i][j][isTrue ? 1 : 0] != -1) return dp[i][j][isTrue ? 1 : 0];

        int ans = 0;
        for(int k = i+1; k <= j-1; k+=2){
            int lt = countWaysMemo(s, i, k-1, true, dp);
            int lf = countWaysMemo(s, i, k-1, false, dp);
            int rt = countWaysMemo(s, k+1, j, true, dp);
            int rf = countWaysMemo(s, k+1, j, false, dp);

            if(s.charAt(k) == '&'){
                if(isTrue) ans += lt * rt;
                else ans += lt * rf + lf * rt + lf * rf;
            }
            else if(s.charAt(k) == '|'){
                if(isTrue) ans += lt * rt + lt * rf + lf * rt;
                else ans += lf * rf;
            }
            else if(s.charAt(k) == '^'){
                if(isTrue) ans += lt * rf + lf * rt;
                else ans += lt * rt + lf * rf;
            }
        }

        return dp[i][j][isTrue ? 1 : 0] = ans;
    }

    // tabulation - Time: O(n^3) Space: O(n^2)
    public int countWaysTab(String s, int n) {
        int[][][] dp = new int[n][n][2];
        for(int i = n-1; i >= 0; i--){
            for(int j = i; j < n; j++){
                if(i == j){
                    if(s.charAt(i) == 'T') dp[i][j][1] = 1;
                    else dp[i][j][0] = 1;
                }
            }
        }

        for(int gap = 2; gap < n; gap+=2){
            for(int i = 0; i < n-gap; i+=2){
                int j = i + gap;
                for(int k = i+1; k <= j-1; k+=2){
                    if(s.charAt(k) == '&'){
                        dp[i][j][1] += dp[i][k-1][1] * dp[k+1][j][1];
                        dp[i][j][0] += dp[i][k-1][0] * dp[k+1][j][0] + dp[i][k-1][1] * dp[k+1][j][0] + dp[i][k-1][0] * dp[k+1][j][1];
                    }
                    else if(s.charAt(k) == '|'){
                        dp[i][j][1] += dp[i][k-1][1] * dp[k+1][j][1] + dp[i][k-1][1] * dp[k+1][j][0] + dp[i][k-1][0] * dp[k+1][j][1];
                        dp[i][j][0] += dp[i][k-1][0] * dp[k+1][j][0];
                    }
                    else if(s.charAt(k) == '^'){
                        dp[i][j][1] += dp[i][k-1][1] * dp[k+1][j][0] + dp[i][k-1][0] * dp[k+1][j][1];
                        dp[i][j][0] += dp[i][k-1][1] * dp[k+1][j][1] + dp[i][k-1][0] * dp[k+1][j][0];
                    }
                }
            }
        }

        return dp[0][n-1][1];

    }

    public static void main(String[] args) {
        Q52 obj = new Q52();
        String s = "T|T&F^T";
        int n = s.length();
        int[][][] dp = new int[n][n][2];
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                for(int k = 0; k < 2; k++){
                    dp[i][j][k] = -1;
                }
            }
        }
        System.out.println(obj.countWays(s, 0, n-1, true));
        System.out.println(obj.countWaysMemo(s, 0, n-1, true, dp));
        System.out.println(obj.countWaysTab(s, n));
    }
}
